src/highlevel_api.jl

```diff
@@ -164,45 +164,72 @@ num_numa_nodes() = count(hwloc_isa(:NUMANode), gettopology())
 """
 Returns a vector containing the sizes of all available L3 caches in Bytes.
 """
-l3cache_sizes() = map(obj->obj.attr.size, Iterators.filter(obj->obj.type_ == :L3Cache, gettopology()))
+l3cache_sizes() = map(obj -> obj.attr.size, Iterators.filter(obj -> obj.type_ == :L3Cache, gettopology()))
 """
 Returns a vector containing the L3 cache line sizes in Bytes.
 """
-l3cache_linesizes() = map(obj->obj.attr.linesize, Iterators.filter(obj->obj.type_ == :L3Cache, gettopology()))
+l3cache_linesizes() = map(obj -> obj.attr.linesize, Iterators.filter(obj -> obj.type_ == :L3Cache, gettopology()))
 
 """
 Returns a vector containing the sizes of all available L2 caches in Bytes.
 """
-l2cache_sizes() = map(obj->obj.attr.size, Iterators.filter(obj->obj.type_ == :L2Cache, gettopology()))
+l2cache_sizes() = map(obj -> obj.attr.size, Iterators.filter(obj -> obj.type_ == :L2Cache, gettopology()))
 """
 Returns a vector containing the L2 cache line sizes in Bytes.
 """
-l2cache_linesizes() = map(obj->obj.attr.linesize, Iterators.filter(obj->obj.type_ == :L2Cache, gettopology()))
+l2cache_linesizes() = map(obj -> obj.attr.linesize, Iterators.filter(obj -> obj.type_ == :L2Cache, gettopology()))
 
 """
 Returns a vector containing the sizes of all available L1 caches in Bytes.
 """
-l1cache_sizes() = map(obj->obj.attr.size, Iterators.filter(obj->obj.type_ == :L1Cache, gettopology()))
+l1cache_sizes() = map(obj -> obj.attr.size, Iterators.filter(obj -> obj.type_ == :L1Cache, gettopology()))
 """
 Returns a vector containing the L1 cache line sizes in Bytes.
 """
-l1cache_linesizes() = map(obj->obj.attr.linesize, Iterators.filter(obj->obj.type_ == :L1Cache, gettopology()))
+l1cache_linesizes() = map(obj -> obj.attr.linesize, Iterators.filter(obj -> obj.type_ == :L1Cache, gettopology()))
 
 """
+    cachesize()
 Returns the L1, L2, and L3 cache sizes in Bytes.
 (Produces a warning if caches of the same level have different sizes across the system.)
 """
 function cachesize()
     allequal = xs -> all(x == first(xs) for x in xs)
+
     l1 = l1cache_sizes()
     l2 = l2cache_sizes()
     l3 = l3cache_sizes()
 
+    isempty(l1) && throw(ErrorException("Your system doesn't seem to have an L1 cache."))
+    isempty(l2) && throw(ErrorException("Your system doesn't seem to have an L2 cache."))
+    isempty(l3) && throw(ErrorException("Your system doesn't seem to have an L3 cache."))
+
     allequal(l1) || (@warn "Not all L1 cache sizes are equal. Consider using `l1cache_sizes()` instead.")
     allequal(l2) || (@warn "Not all L2 cache sizes are equal. Consider using `l2cache_sizes()` instead.")
     allequal(l3) || (@warn "Not all L3 cache sizes are equal. Consider using `l3cache_sizes()` instead.")
 
-    return (L1=first(l1), L2=first(l2), L3=first(l3))
+    return (L1=maximum(l1), L2=maximum(l2), L3=maximum(l3))
+end
+
+"""
+    cachesize(x::Symbol) -> size
+Returns the size of the specified cache type. Allowed inputs are `:L1`, `:L2`, and `:L3`.
```

Comment:
Why is `:I1` not allowed? Isn't that introduced earlier?

Why use `:L1` instead of `:L1Cache` here? Isn't `:L1Cache` used above?

Reply:
> Why is `:I1` not allowed? Isn't that introduced earlier?

I introduced `cachesize` (and `cachelinesize`) as high-level functions for grabbing information about common caches. It's never going to be "general" because a system could have whatever other caches (e.g., perhaps, an L4 cache) and we're always only supporting a finite list. The user can always fall back to querying the size manually (by iterating the topology tree) or by investigating the output of `topology()` of course. So, we could add `:I1` but it's an instruction cache and IMO typically not of great importance. What do you think?



> Why use `:L1` instead of `:L1Cache` here? Isn't `:L1Cache` used above?

The hwloc objects in the topology are called `:L1Cache`, yes. But I aligned it with the output of `cachesize()` (without arguments) which returns, e.g., `(L1 = 32768, L2 = 1048576, L3 = 28835840)` which is arguably nicer than `(L1Cache = 32768, L2Cache = 1048576, L3Cache = 28835840)`. Also, `:L1` is probably more intuitive than `:L1Cache`. But I don't have a strong opinion here.

Reply:
Ok.

Reply:
Ok.